### PR TITLE
handle duplicate definitions in generate_repo_config

### LINKED ISSUE #968: duplicate go_repository rules are now rejected
### What version of gazelle are you using?

0.22.2

### What version of rules_go are you using?

0.25.0

### What version of Bazel are you using?

3.6

As of some recent gazelle version (probably whatever picked up https://github.com/bazelbuild/bazel-gazelle/pull/855), duplicate go_repository rules are rejected in at least some combinations that were previously accepted.

The error will be reported as "multiple rules have the name %s". There turns out to be a range of scenarios where this can occur which are arguably incorrect. I have only partially debugged it so far, so I'm going to be a bit vague about what those scenarios are, but it includes things like:

- The WORKSPACE file includes both repositories defined in external _deps() functions and in go_dependencies.bzl generated by gazelle update-repos
- Some of those repository names match, and the ordering in WORKSPACE has been arranged so that gazelle's definitions override the previous ones.
- gazelle is told to only update go_dependencies.bzl

I believe this scenario should be accepted by gazelle, and currently (at least in some combinations, details TBD) it is rejected. Another similar scenario is where a `# gazelle:repository ` line is overriding something generated in go_dependencies.bzl.

I suspect, but have not yet proved, that this is happening because of the collapsing of definitions into @bazel_gazelle_go_repository_config losing information. Certainly in the cases I'm working on debugging there are two entries in that config which are literal duplicates of each other. My working hypothesis is that generate_repo_config is misinterpreting these scenarios as duplicate rules.

I'll do some more debugging tomorrow, but I figured I'd dump my work so far into this issue in case the problem is immediately obvious to you.

### PR DESCRIPTION
**Pull request by tfrench-uber**
This is a commandeering of @asuffield's PR https://github.com/bazelbuild/bazel-gazelle/pull/979 with some minor changes:

From 979:

What type of PR is this?

Bug fix

What package or component does this PR mostly affect?

go_repository

What does this PR do? Why is it needed?

ListRepositories currently attempts to merge duplicates, but has several bugs in this code. Refactored the whole thing to have no duplicates in the output and to match bazel's own logic for picking the winner. In scenarios where there are both gazelle directives and rules, we prefer the gazelle directive. When multiple gazelle directives are found, we pick the last one; to make this work, ListRepositories is adjusted to maintain the ordering of rules.

This notably fixes the way generate_repo_config interacts with overrides between multiple # gazelle:repository_macro definitions, which currently breaks. It likely also fixes other corner cases where there are multiple definitions with different importpath, which would confound both update-repos and fix-update, although I don't have a reproduction case.

The test added to generate_repo_config_test.go is a minimal scenario that fails on the latest release and works after this change.

Which issues(s) does this PR fix?

Fixes #968

### CODE REVIEW

File: cmd/generate_repo_config/generate_repo_config_test.go [hunk 1]
```diff
@@ -104,3 +104,73 @@ go_repository(
 		},
 	})
 }
+
+func TestDuplicateRepo(t *testing.T) {
```
Comment (blico) [inline cmd/generate_repo_config/generate_repo_config_test.go:108]: NIT: This test and the one above it seem to have a very similar structure, they could probably be merged into a table test.

File: cmd/generate_repo_config/generate_repo_config.go [hunk 2]
```diff
@@ -104,8 +111,12 @@ func generateRepoConfig(configDest, configSource string) ([]string, error) {
...
 		for _, d := range r.Directives {
 			// skip repository_macro directives, because for the repo config we flatten
-			// macros into one file
-			if d.Key != "repository_macro" {
+			// macros into one file. skip repository directives because they are merged
+			// into repos already.
+			switch d.Key {
+			case _repositoryMacroKey, _repositoryKey:
+				continue
+			default:
```
Comment (blico) [inline cmd/generate_repo_config/generate_repo_config.go:119]: Right now this case will never be hit because `repositoryMacroKey` and `repositoryKey` are the only workspace directives. Maybe we should assume that `ListRepositories` will process all workspace directives so there is no need to add them into the merged file? That way we can remove all of this logic.

**Review by blico — COMMENTED** [review]
Looks good other than some minor feedback.

File: repo/repo.go [hunk 3]
```diff
@@ -66,98 +67,131 @@ func FindExternalRepo(repoRoot, name string) (string, error) {
...
+	repoRoot string
+
+	repoFileMap  map[string]*rule.File // repo rule name => file that contains repo
+	repoIndexMap map[string]int        // repo rule name => index of rule in "repos" slice
+	visited      map[macroKey]struct{}
+}
+
+func isFromDirective(repo *rule.Rule) bool {
+	b, ok := repo.PrivateAttr(config.GazelleFromDirectiveKey).(bool)
+	if !ok {
```
[the remark this answers is not in the document]
Comment (tfrench-uber) [reply repo/repo.go]: Good point! updated

File: cmd/generate_repo_config/generate_repo_config.go [hunk 4]
```diff
@@ -39,9 +39,16 @@ import (
 	"github.com/bazelbuild/bazel-gazelle/rule"
 )
 
+const (
+	_repositoryMacroKey  = "repository_macro"
```
[the remark this answers is not in the document]
Comment (tfrench-uber) [reply cmd/generate_repo_config/generate_repo_config.go]: Agreed - made the change
